Retry http 502, 503, 504 status codes. Fixes #25

janpieterz opened the pull request:
Tried to get a unit test going but the only thing I could think of (using httpbin) was overwriting debug and checking that statement.

stephenmichaelf commented on tool.ts:
```diff
@@ -218,10 +222,20 @@ export async function downloadTool(url: string, fileName?: string): Promise<stri
-            // TODO: retries
+            
             tl.debug('downloading');
+            const  statusCodesToRetry = [502, 503, 504];
```
Add type, extra space between type and variable, can use these for values:

502 - httpm.HttpCodes.BadGateway
503 - httpm.HttpCodes.ServiceUnavailable
504 - httpm.HttpCodes.GatewayTimeout

stephenmichaelf commented on tool.ts:
```diff
@@ -218,10 +222,20 @@ export async function downloadTool(url: string, fileName?: string): Promise<stri
+            
             tl.debug('downloading');
+            const  statusCodesToRetry = [502, 503, 504];
+            let retryCount = 0;
```
Add type

stephenmichaelf commented on tool.ts:
```diff
@@ -218,10 +222,20 @@ export async function downloadTool(url: string, fileName?: string): Promise<stri
+            let retryCount = 0;
             let response: httpm.HttpClientResponse = await http.get(url);
+
+            while(retryCount < 2 && statusCodesToRetry.indexOf(response.message.statusCode) > -1) {
```
Extract 2 to variable, maxRetries or something along those lines.

stephenmichaelf commented on tool.ts:
```diff
@@ -218,10 +222,20 @@ export async function downloadTool(url: string, fileName?: string): Promise<stri
             let response: httpm.HttpClientResponse = await http.get(url);
+
+            while(retryCount < 2 && statusCodesToRetry.indexOf(response.message.statusCode) > -1) {
+                tl.debug(`Download attempt "${retryCount + 1}" failed.`);
```
Add response.message.statusCode to debug message.

stephenmichaelf commented on tool.ts:
```diff
@@ -218,10 +222,20 @@ export async function downloadTool(url: string, fileName?: string): Promise<stri
+                tl.debug(`Download attempt "${retryCount + 1}" failed.`);
+                retryCount += 1;
+                await delay(1000);
+                tl.debug(`Downloading attempt "${retryCount + 1}"`)
```
Download attempt X of max Y?

janpieterz commented:
@stephenmichaelf implemented your changes, made sense!

Just to be sure we're not missing it, the code ran fine but I haven't actually verified the retry working. Any thoughts there? Otherwise I'll setup a quick server here that responds on the third request only.

janpieterz commented:
Got what you meant with types, thought you meant the http code types, fixed. 
Removed need for +1.
Added unit test (commented), can remove if you prefer.
Got a quick local server (node, express) and confirmed:

```js
const express = require("express");
const app = express();

let counting = 0;

app.get("/", (req, res) => {
  counting++;
  if (counting % 2 == 0) {
    console.log("Sending success");
    res.send("Hello world!");
  } else {
    console.log("Sending failure");
    res.sendStatus(502);
  }
});

app.listen(3000, () => console.log("Example app listening on port 3000!"));
```

janpieterz commented:
@stephenmichaelf Going to be able to put some more work into this end of this week. Any chance for a review?

stephenmichaelf commented:
@janpieterz Were you able to get some more work done on this? Thanks!

janpieterz commented:
Think the work was done, unless I missed feedback somewhere?

> On 12 Oct 2018, at 14:01, Stephen Franceschelli <notifications@github.com> wrote:
> 
> @janpieterz Were you able to get some more work done on this? Thanks!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

damccorm commented:
@janpieterz this generally looks good - thanks for the contribution and sorry about the delay! I think having unit testing makes sense, it probably would be better to mock the http calls though. [Nock](https://www.npmjs.com/package/nock) is a really good tool for this, I've used it in some of our other libraries for pretty similar use cases, there's some examples in ```tests/units``` of this library

Would you be willing to add some mocked tests? I'm also definitely open to other methods of doing this, I'd just rather avoid having to set up a server to run tests. Let me know if I can help brainstorm/with Nock!

janpieterz commented:
@damccorm No worries, happy to get this merged.

Not familiar with Nock but looks good. Don't have a lot of time the upcoming weeks, so maybe someone would be willing to integrate Nock first? I'd be happy to refactor the test to use Nock.
Otherwise I'm thinking somewhere around the end of the month.

damccorm commented on tool.ts:
```diff
@@ -218,10 +222,21 @@ export async function downloadTool(url: string, fileName?: string): Promise<stri
+                tl.debug(`Download attempt "${retryCount}" of "${maxRetries}" failed with status code "${response.message.statusCode}".`);
+                retryCount += 1;
+                await delay(1000);
+                tl.debug(`Downloading attempt "${retryCount}" of "${maxRetries}"`)
```
Nit: semicolon missing

janpieterz commented:
@damccorm Sweet, didn't see those upstream changes. Have merged upstream master to branch, fixed nit and added your unit tests + confirmed here. 

FYI, I did have to change make.js to actually run those unit tests (did not commit)
![image](https://user-images.githubusercontent.com/383634/49656880-b49e7500-fa3e-11e8-95c9-8415466b2ac7.png)

damccorm commented:
This looks great! One last thing: would you mind bumping the patch version in ```package.json``` and ```package-lock.json``` to 0.10.2 so we can release?

Also, concerning the make.js, if you want to commit your change that would be great! Though, maybe we can just replace the block
```
    console.log("-------Unit Tests-------");
    run('tsc -p ./test/units --outDir ' + unitPath, true);
```

with a call to ```target.units()``` for better modularity?

janpieterz commented:
@damccorm Sorted!
